sendrecv/gst-rust/src/main.rs

```diff
@@ -0,0 +1,480 @@
+        Ok(())
+    } {
+        Ok(()) => return None,
+        Err(err) => panic!("Error adding pad with caps {} {:?}", name, err),
+    };
+}
+
+fn on_incoming_stream(values: &[glib::Value], pipe: &gst::Element) -> Option<glib::Value> {
+    let webrtc = values[0].get::<gst::Element>().expect("Invalid argument");
+    let decodebin = gst::ElementFactory::make("decodebin", None).unwrap();
+    let pipe_clone = pipe.clone();
+    decodebin
+        .connect("pad-added", false, move |values| {
+            on_incoming_decodebin_stream(values, &pipe_clone)
+        })
+        .unwrap();
+    pipe.clone()
+        .dynamic_cast::<gst::Bin>()
+        .unwrap()
+        .add(&decodebin)
+        .unwrap();
+    decodebin.sync_state_with_parent().unwrap();
+    webrtc.link(&decodebin).unwrap();
+    None
+}
+
+fn send_ice_candidate_message(
+    app_control: &Arc<Mutex<AppControl>>,
+    values: &[glib::Value],
+) -> Option<glib::Value> {
+    let app_control = app_control.lock().unwrap();
+    if app_control.app_state < AppState::PeerCallNegotiating {
+        panic!("Can't send ICE, not in call");
+    }
+
+    let _webrtc = values[0].get::<gst::Element>().expect("Invalid argument");
+    let mlineindex = values[1].get::<u32>().expect("Invalid argument");
+    let candidate = values[2].get::<String>().expect("Invalid argument");
+    let message = json!({
+          "ice": {
+            "candidate": candidate,
+            "sdpMLineIndex": mlineindex,
+          }
+        });
+    app_control.ws_sender.send(message.to_string()).unwrap();
+    None
+}
+
+fn start_pipeline(app_control: &Arc<Mutex<AppControl>>) -> Result<gst::Element, Error> {
+    let pipe = gst::parse_launch(&format!(
+        "webrtcbin name=sendrecv {}
+        videotestsrc pattern=ball ! videoconvert ! queue ! vp8enc deadline=1 ! rtpvp8pay !
+        queue ! {}96 ! sendrecv.
+        audiotestsrc wave=red-noise ! audioconvert ! audioresample ! queue ! opusenc ! rtpopuspay !
+        queue ! {}97 ! sendrecv.
+        ",
+        STUN_SERVER, RTP_CAPS_VP8, RTP_CAPS_OPUS
+    ))?;
+
+    let webrtc = pipe.clone()
+        .dynamic_cast::<gst::Bin>()
+        .unwrap()
+        .get_by_name("sendrecv")
+        .unwrap();
+    let app_control_clone = app_control.clone();
+    webrtc.connect("on-negotiation-needed", false, move |values| {
+        on_negotiation_needed(&app_control_clone, values)
+    })?;
+
+    let app_control_clone = app_control.clone();
+    webrtc.connect("on-ice-candidate", false, move |values| {
+        send_ice_candidate_message(&app_control_clone, values)
+    })?;
+
+    let pipe_clone = pipe.clone();
+    // TODO: replace with webrtc.connect_pad_added
+    webrtc.connect("pad-added", false, move |values| {
+        on_incoming_stream(values, &pipe_clone)
+    })?;
+
+    pipe.set_state(gst::State::Playing).into_result()?;
+
+    return Ok(webrtc);
+}
+
+struct WsClient {
+    webrtc: Option<gst::Element>,
+    app_control: Arc<Mutex<AppControl>>,
```

Comment:
I would do this as follow
```rust
#[derive(Clone)]
struct AppControl(Arc<Mutex<AppControlInner>>);`

struct AppControlInner {
   [the fields from below]
}
```

And then pass a `&AppControl` to all the functions... or maybe even make all the functions you have above part of an `impl AppControl` block

Reply:
Also should there be maybe only one struct instead of a separate `WsClient` that only has the element in addition?